### PR TITLE
[WIP] Ceil, Floor, Log, Round, Sqrt layers added

### LINKED ISSUE #16749: Add layers implementation for Object detection & Image Segmentation model at ONNX model Zoo
##### System information (version)

- OpenCV => 3.4.9-dev
- Operating System / Platform => Ubuntu

##### Detailed description

Layer implementation of few layers that are needed to load Object detection and Image segmentation model from ONNX model Zoo are not present. Some of them includes-
`Ceil`
`Floor`
`Log`
`Round `
`Sqrt`

**Few examples** - 
[ObjectDetection&ImageSegmentation/yolov3/yolov3](https://github.com/onnx/models/blob/master/vision/object_detection_segmentation/yolov3)
`Ceil`
[ObjectDetection&ImageSegmentation/faster_rcnn_R_50_FPN_1x/faster_rcnn_R_50_FPN_1x](https://github.com/onnx/models/blob/master/vision/object_detection_segmentation/faster-rcnn)
`Floor`
`Log`
`Sqrt`

##### Steps to reproduce
Not an error but feature request.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution

 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [ ] There is reproducer code and related data files: videos, images, onnx, etc

### PR DESCRIPTION
**Pull request by ashishkrshrivastava**
resolves #16749 

**Merge with extra**: https://github.com/opencv/opencv_extra/pull/723

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### This PR adds implementation of 

- [x] `Ceil Layer`
- [x] `Floor Layer`
- [x] `Log Layer`
- [x] `Round Layer`
- [x] `Sqrt Layer`
- [x] `Exp Layer`
- [x] `Not Layer`
- [x] `ReduceMin` (only for constant because it is working on constants in fasterrcnn and maskrcnn )
- [x] `NonZero` (only for constant blob because it is working on constants in fasterrcnn and maskrcnn)
- [x] `Equal`
- [x] `Greater`
- [x] `Less`
- [x] `And`
- [ ] `Gather`
- [ ] `TopK`
<cut/>

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.1.0:16.04
build_image:Custom Win=openvino-2020.1.0
build_image:Custom Mac=openvino-2020.1.0

Xtest_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by ashishkrshrivastava** [comment]
@alalek @dkurt I am not sure why it is failing here. these changes are working perfectly fine locally. Any idea how to tackle this?

**Comment by alalek** [comment]
`std::round` is C++ feature: https://en.cppreference.com/w/cpp/numeric/math/round
OpenCV 3.4 should use C++98 by default.

Failed builders have GCC 4.x.

**Comment by ashishkrshrivastava** [comment]
@alalek thanks.

**Comment by ashishkrshrivastava** [comment]
@alalek , is there any documentation where I can get to know about the configuration of builders like what version of gcc or C++ custom builder use ?
changes are only getting failed for custom builders.

**Comment by alalek** [comment]
> version of gcc

Check "CMake" step => "stdio" log.

> failed for custom builders

These builders are configured to build OpenCV with OpenVINO's Inference engine (Open Source version is here: https://github.com/opencv/dldt/releases/tag/2020.1).

General information about IE and how to build OpenCV with IE locally: https://github.com/opencv/opencv/wiki/Intel%27s-Deep-Learning-Inference-Engine-backend#build-opencv-from-source
This should work well if you using Ubuntu 16.04 / 18.04.

> supportBackend

You should tune implementation details through `.supportBackend()` method. Simplest way is to disable support for both `DNN_BACKEND_INFERENCE_ENGINE_NN_BUILDER_2019` and `DNN_BACKEND_INFERENCE_ENGINE_NGRAPH` backend. These implementations can be added later (if possible).

---

General information about CI options: https://github.com/opencv/opencv/wiki/CI-configuration

**Comment by ashishkrshrivastava** [comment]
@alalek Thank you so much.

**Comment by dkurt** [comment]
Please see the comment https://github.com/opencv/opencv/issues/16749#issuecomment-596207947. I think that adding so atomic operations is not a right way to enable object detection networks. They won't cover all the operations from the models and we need more fundamental solution. Can you specify origin frameworks for YOLOv3 and Faster R-CNN models in ONNX format?

**Comment by ashishkrshrivastava** [comment]
@dkurt , all of object detection models are originated from keras2onnx. They are written in keras.

**Comment by alalek** [comment]
> I think that adding so atomic operations is not a right way to enable object detection networks.

It can be useful, for example, for debugging purposes of graph transformations.

**Comment by ashishkrshrivastava** [comment]
@dkurt 

> we need more fundamental solution. 

I have added explanation in [#16749 comment](https://github.com/opencv/opencv/issues/16749#issuecomment-597376482) . Please share your view.

**Comment by ashishkrshrivastava** [comment]
@alalek , necessary lines of code is already added to PR description to allow custom build then why it is not getting build. Am I missing something ?

**Comment by ashishkrshrivastava** [comment]
@alalek , sorry opencv/opencv_extra#723 had merge conflict yesteday, I have resolved them. can you please let build on custom builders.

**Comment by asmorkalov** [comment]
I re-triggered all builds, include custom one.

**Comment by ashishkrshrivastava** [comment]
> I re-triggered all builds, include custom one.

Thank you so much @asmorkalov . They all are build successfully.

**Comment by dkurt** [comment]
> resolves #16749

Please specify if this PR enables one of the models mentioned in the issue

**Comment by ashishkrshrivastava** [comment]
@dkurt Actually I created the issue requesting layer implementation, not model. But as you said that we must focus on models not operators or layers, making models work needs several fusion and layer implementation. It would take time. So if you want me to make model work, please let the PR open. I will complete that. Otherwise this PR can be merged for now since layer implementation is already there and I will work on models separately.

**Comment by ashishkrshrivastava** [comment]
@alalek , I am adding a layer which needs changes in `usr/local/include/opencv2/dnn/all_layers.hpp` and using those changes more changes are required in `modules/dnn/src/layers/pooling_layer.cpp` but builder can only find changes in pooling_layer.cpp not in all_layers.hpp. 
I am using the script 
```
cd ~/opencv/build
cmake -D CMAKE_BUILD_TYPE=DEBUG \
	-D CMAKE_INSTALL_PREFIX=/usr/local \
	-D INSTALL_PYTHON_EXAMPLES=ON \
	-D INSTALL_C_EXAMPLES=OFF \
	-D OPENCV_ENABLE_NONFREE=ON \
	-D OPENCV_EXTRA_MODULES_PATH=~/opencv_contrib/modules \
	-D PYTHON_EXECUTABLE=~/.virtualenvs/cv/bin/python3 \
	-D BUILD_EXAMPLES=ON ..

make -j4
sudo make install
sudo ldconfig
ls /usr/local/lib/python3.6/site-packages/cv2/python-3.6
cd /usr/local/lib/python3.6/site-packages/cv2/python-3.6
sudo mv cv2.cpython-36m-x86_64-linux-gnu.so  cv2.so
cd ~/.virtualenvs/cv/lib/python3.6/site-packages/
ln -s /usr/local/lib/python3.6/site-packages/cv2/python-3.6/cv2.so cv2.so

```
Am i missing something ?

I am sorry to ask this question here.

**Comment by alalek** [comment]
> usr/local/include

out of source tree.

**Comment by ashishkrshrivastava** [comment]
> out of source tree.

ohh.. need to add changes here `/opencv/modules/dnn/include/opencv2/dnn/all_layers.hpp`
Thankyou so much.

**Comment by ashishkrshrivastava** [comment]
I think these changes are not relevant anymore. Closing it for now. I would definitely work on it if find any issue regarding this in future.
